api/src/services/authorization.js

```diff
@@ -0,0 +1,225 @@
+
+  return Promise
+    .all([
+      db.medic.query('medic/contacts_by_depth', { keys: keys }),
+      db.medic.query('medic-tombstone/contacts_by_depth', { keys: keys }),
+    ])
+    .then(resultSets => {
+      const subjectIds = [];
+
+      resultSets.forEach((resultSet, tombstone) => {
+        resultSet.rows.forEach(row => {
+          subjectIds.push( tombstone ? tombstoneUtils.extractDocId(row.id) : row.id );
+          if (row.value) {
+            subjectIds.push(row.value);
+          }
+        });
+      });
+
+      subjectIds.push(ALL_KEY);
+      if (hasAccessToUnassignedDocs(userCtx)) {
+        subjectIds.push(UNASSIGNED_KEY);
+      }
+
+      return subjectIds;
+    });
+};
+
+/**
+ * Method to ensure users don't see reports submitted by their boss about the user
+ */
+const isSensitive = function(userCtx, subject, submitter, allowedSubmitter) {
+  if (!subject || !submitter) {
+    return false;
+  }
+
+  if (subject !== userCtx.contact_id && subject !== userCtx.facility_id) {
+    return false;
+  }
+
+  return !allowedSubmitter;
+};
+
+const getValidatedDocIds = (subjectIds, userCtx) => {
+  return Promise
+    .all([
+      db.medic.query('medic/docs_by_replication_key', { keys: subjectIds }),
+      db.medic.query('medic-tombstone/docs_by_replication_key', { keys: subjectIds })
+    ])
+    .then(resultSets => {
+      const validatedIds = ['_design/medic-client', 'org.couchdb.user:' + userCtx.name];
+
+      resultSets.forEach(resultSet => {
+        resultSet.rows.forEach(row => {
+          if (!isSensitive(userCtx, row.key, row.value.submitter, subjectIds.indexOf(row.value.submitter) !== -1)) {
+            validatedIds.push(row.id);
+          }
+        });
+      });
+
+      return validatedIds;
+    });
+};
+
+const getViewResults = (doc) => {
+  return {
+    contactsByDepth: viewMapUtils.getViewMapFn('medic', 'contacts_by_depth', true)(doc),
+    replicationKey: viewMapUtils.getViewMapFn('medic', 'docs_by_replication_key')(doc)
+  };
+};
+
+const allowedChange = (feed, changeObj) => {
+  const userOpts = _.pick(feed, 'userCtx', 'subjectIds', 'depth');
+  return allowedDoc(changeObj.change.doc, userOpts, changeObj.viewResults);
+};
+
+const isAuthChange = (feed, changeObj) => {
+  if (changeObj.change.id !== 'org.couchdb.user:' + feed.userCtx.name) {
+    return false;
+  }
+
+  if (feed.userCtx.contact_id !== changeObj.change.doc.contact_id ||
+      feed.userCtx.facility_id !== changeObj.change.doc.facility_id) {
+    return true;
+  }
+
+  return false;
+};
+
+module.exports = {
+  allowedChange: allowedChange,
+  isAuthChange: isAuthChange,
+  allowedDoc: allowedDoc,
+  getDepth: getDepth,
+  getViewResults: getViewResults,
+  getSubjectIds: getSubjectIds,
+  getValidatedDocIds: getValidatedDocIds,
+};
+
+// used for testing
+if (process.env.UNIT_TEST_ENV) {
+  _.extend(module.exports, {
+    _tombstoneUtils: tombstoneUtils,
+    _viewMapUtils: viewMapUtils
+  });
```

Comment:
I think this isn't necessary - if the unit test requires the shared lib then mocks the functions it should work.

Reply:
You're right. It was necessary for tombstoneUtils, because, in it's old form, it was initialized when it was required. I have refactored that part of the lib, so initializing is not necessary anymore.